kernel/src/paging.c

```diff
@@ -0,0 +1,170 @@
+
+    for (int i = 0; i < 512; i++) {
+        // apply flags and update address in new page table
+        new_addr[i] = current_addr + i * delta + g_pagetable_flags_default;
+    }
+
+    return create_table_entry(new_addr);
+}
+
+// allocates a new table if needed
+uint64_t alloc_missing(uint64_t table_entry) {
+    if (table_entry & f_paging_present) {
+        return table_entry;
+    }
+
+    return create_table_entry(alloc_page());
+}
+
+typedef struct {
+    uint64_t virtaddr_low;
+    uint64_t virtaddr_high;
+    uint64_t physoffset;
+    uint64_t flags;
+} map_params;
+
+void rec_setpage(uint64_t* table, uint64_t tableaddr, uint8_t depth, map_params* params) {
+    // align table address with virutal address
+    // this is the same as finding the lower bound of a table
+    if (tableaddr < params->virtaddr_low) {
+        tableaddr += params->virtaddr_low & pagingmasks[depth];
+    }
+
+    uint64_t delta = 1L << (3 + 9 * depth);
+
+    // go through all relevant tables
+    int low = get_table(tableaddr, depth);
+    while (low < 512 && tableaddr <= params->virtaddr_high) {
+        const uint64_t physaddr = tableaddr + params->physoffset;
+
+        const bool filled_page = (tableaddr + delta) <= params->virtaddr_high;
+        const bool is_large = table[low] & f_paging_large;
+        const bool is_present = table[low] & f_paging_present;
+        const bool P1 = depth <= 1, P2 = depth == 2;
+
+        uint64_t applied_flags = params->flags & ~f_paging_ptrmask;
+
+        // set_value maps pagetable adresses and goes to the next iteration
+        if (P1) goto set_value;
+
+        if (P2 && filled_page) {
+            if (is_large || !is_present) goto set_value;
+
+            // already subdivided
+            rec_setpage((uint64_t*)(table[low] & f_paging_ptrmask), tableaddr, depth - 1, params);
+            goto next;
+        }
+
+        // table entry must be split into subsegments
+        table[low] = alloc_missing(table[low]);
+        table[low] = subdivide_large(table[low], depth);
+
+        rec_setpage((uint64_t*)(table[low] & f_paging_ptrmask), tableaddr, depth - 1, params);
+        goto next;
+
+    set_value:
+
+        if (!P1) applied_flags |= f_paging_large;
+
+        table[low] = (physaddr & f_paging_ptrmask) | applied_flags;
+
+    next:
+        // incremenet current page
+        tableaddr += delta;
+        low++;
+    }
+}
+
+void paging_map(uint64_t* table, uint64_t low, uint64_t high, uint64_t phys_low, uint64_t flags) {
+    map_params params = {low & f_paging_ptrmask,
+                         high & f_paging_ptrmask,
+                         (phys_low - low) & f_paging_ptrmask,
+                         flags};
+
+    rec_setpage(table, 0, 4, &params);
+}
+
+void paging_map_identity(uint64_t* table, uint64_t low, uint64_t high) {
+    paging_map(table, low, high, low, g_pagetable_flags_default);
+}
+
+uint64_t* paging_fetch(uint64_t* table, uint64_t virtaddr) {
```

Comment:
This function can return a NULL pointer, there should probably be some sort of explanation as to why and when this happens